### PR TITLE
Block Machine: Handle case where number of rows is just enough

### PR DESCRIPTION
**Pull request by georgwiese**
Before this PR, we used to panic if the number of rows in a block machine was just enough to fit the needed number of blocks. This PR fixes it.

### CODE REVIEW

File: parser/src/lib.rs [hunk 1]
```diff
@@ -256,6 +256,7 @@ mod test {
                     };
                     eprint!("\t{sign}{change}");
                 }
+                eprintln!("The following string was re-parsed:\n{orig_asm_to_string}");
```
Comment (georgwiese) [inline parser/src/lib.rs:259]: Somehow, the newly added test fails the re-parse test! It seems to forget the `Main` min and max degree:

```
The following string was re-parsed:
use std::machines::binary::ByteBinary as ByteBinary;
use std::machines::large_field::binary::Binary as Binary;
machine Main {
    reg pc[@pc];
    reg X0[<=];
    reg X1[<=];
    reg X2[<=];
    reg A;
    ByteBinary byte_binary;
    Binary binary(byte_binary, 8, 8);
    instr and X0, X1 -> X2 link ~> X2 = binary.and(X0, X1){  }
    function main {
    A <== and(2863311530, 2863311530);
    A <== and(4294967295, 4294967295);
    return;
    }
}

thread 'test::parse_write_reparse_asm' panicked at parser/src/lib.rs:260:17:
parsed and re-parsed ASTs differ for file: /Users/georg/coding/powdr/test_data/asm/block_machine_exact_number_of_rows.asm
```

**Comment by leonardoalt** [comment]
Test fails

File: executor/src/witgen/machines/block_machine.rs [hunk 2]
```diff
@@ -279,7 +279,13 @@ impl<'a, T: FieldElement> Machine<'a, T> for BlockMachine<'a, T> {
                  This might violate some internal constraints."
             );
         }
-        self.degree = compute_size_and_log(&self.name, self.data.len(), self.degree_range);
+        self.degree = compute_size_and_log(
+            &self.name,
+            // At this point, the still contains the dummy block, which will be removed below.
```
Comment (leonardoalt) [inline executor/src/witgen/machines/block_machine.rs:284]: The?

File: executor/src/witgen/machines/block_machine.rs [hunk 3]
```diff
@@ -279,7 +279,13 @@ impl<'a, T: FieldElement> Machine<'a, T> for BlockMachine<'a, T> {
...
+            // Therefore, we subtract the block size here.
```
Comment (leonardoalt) [inline executor/src/witgen/machines/block_machine.rs:285]: This feels kinda hacky..

File: parser/src/lib.rs (hunk 1 above)
Comment (leonardoalt) [reply parser/src/lib.rs:259]: Ohhh

Comment (georgwiese) [reply parser/src/lib.rs:259]: Fixed thanks to #2079 😍

File: executor/src/witgen/machines/block_machine.rs (hunk 3 above)
Comment (georgwiese) [reply executor/src/witgen/machines/block_machine.rs:176]: Hmm, so I think the hacky part is that the data has a block that will be removed later, right? Which was the case before this PR?

I tried refactoring it, but it's kinda complicated. LMK if you have a suggestion.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**